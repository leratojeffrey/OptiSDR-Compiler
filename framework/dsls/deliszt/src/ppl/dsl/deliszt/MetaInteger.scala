package ppl.dsl.deliszt

/**
 * author: Michael Wu (mikemwu@stanford.edu)
 * last modified: 04/28/2011
 *
 * Pervasive Parallelism Laboratory (PPL)
 * Stanford University
 */

/*
 * MetaInteger.scala
 *
 * To change this template, choose Tools | Template Manager
 * and open the template in the editor.
 */

abstract class IntM {
  type Suc <: IntM
  type Add[N <: IntM] <: IntM
}

object EnsureSize {
  implicit def EnsureSizeZero[T <: IntM] =
  new EnsureSize[Zero,Succ[T]] {
    def get(n : Int) : Int = n
  }

  implicit def EnsureSizeSucc[T <: IntM, TT <: IntM](implicit fn : EnsureSize[TT,T]) =
  new EnsureSize[Succ[TT],Succ[T]] {
    def get(n : Int) : Int = fn.get(n+1)
  }
}

//meta-function(?)(what is the term for this thing?)
//if IDX is less than SIZE, then an implicit object of type EnsureSize[IDX,SIZE] will
//exist with a idx function that returns the number IDX as an integer
abstract class EnsureSize[TT <: IntM,T <: IntM] {
  def get(n : Int) : Int
  def idx = get(0)
}

object MVal {
  implicit val MValZero = new MVal[Zero] {
    val value = 0
  }

  implicit def MValSucc[T <: IntM](implicit f: MVal[T]) = new MVal[Succ[T]] {
    val value = f.value + 1
  }
}

abstract class MVal[T] {
  val value: Int
}

class Zero extends IntM {
  type Suc = Succ[Zero]
  type Add[N <: IntM] = N
}

class Succ[N <: IntM] extends IntM {
  type Suc = Succ[Succ[N]]
  type Add[M <: IntM] = (N#Add[M])#Suc
}

object MetaInteger {
  type +[N1 <: IntM, N2 <: IntM] = N1#Add[N2]

  //Scala doesn't provide meta-integers like C++'s typename<int N>
  //so we have to define some integers here
  //this is a simplification of:
  //http://michid.wordpress.com/code/meta-programming-with-scala-part-i-addition/
  //somelike like Church encoding but the numbers are themselves objects and not
  //functions for simplicity since we don't need to define addition on them
  //as far I can tell ... if necessary I suppose we can define multiplcation
  //on these if we needed numbers higher than 9
  type _0 = Zero
  type _1 = Succ[_0]
  type _2 = Succ[_1]
  type _3 = Succ[_2]
  type _4 = Succ[_3]
  type _5 = Succ[_4]
  type _6 = Succ[_5]
  type _7 = Succ[_6]
  type _8 = Succ[_7]
  type _9 = Succ[_8]
  type _10 = Succ[_9]
  type _11 = Succ[_10]
  type _12 = Succ[_11]
  type _13 = Succ[_12]
  type _14 = Succ[_13]
  type _15 = Succ[_14]
  type _16 = Succ[_15]
  type _17 = Succ[_16]
  type _18 = Succ[_17]
  type _19 = Succ[_18]
  type _20 = Succ[_19]
  type _21 = Succ[_20]
  type _22 = Succ[_21]
  
  //for convience we provide objects representing each number as well
  //so you can avoid passing parameter arguments
  val _0 : _0 = new _0
  val _1 : _1 = new _1
  val _2 : _2 = new _2
  val _3 : _3 = new _3
  val _4 : _4 = new _4
  val _5 : _5 = new _5
  val _6 : _6 = new _6
  val _7 : _7 = new _7
  val _8 : _8 = new _8
  val _9 : _9 = new _9
  val _10 : _10 = new _10
  val _11 : _11 = new _11
  val _12 : _12 = new _12
  val _13 : _13 = new _13
  val _14 : _14 = new _14
  val _15 : _15 = new _15
  val _16 : _16 = new _16
  val _17 : _17 = new _17
  val _18 : _18 = new _18
  val _19 : _19 = new _19
  val _20 : _20 = new _20
  val _21 : _21 = new _21
  val _22 : _22 = new _22

  def MIntDepth[T <: IntM]()(implicit n: MVal[T]) = n.value
  def MIntDepth[T <: IntM](m : T)(implicit n: MVal[T]) = n.value
}
