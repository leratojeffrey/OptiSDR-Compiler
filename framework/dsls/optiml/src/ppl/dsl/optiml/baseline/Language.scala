package ppl.dsl.optiml.baseline

object LanguageDefs {
  val INITIAL_SEED = 100
  var randRef = new scala.util.Random(INITIAL_SEED)
  var _time = 0L
  
  def random(): Double = randRef.nextDouble
    
  def abs(d: Double) = implicitly[Numeric[Double]].abs(d) //java.lang.Math.abs(d)
  
  def sum[B:Numeric:Manifest](st: Int, en: Int)(f: Int => B) = {
    var acc = f(st)
    for (i <- st+1 until en) {
      acc = implicitly[Numeric[B]].plus(acc,f(i))
    }
    acc
  }
  
  def tic() {
    _time = System.currentTimeMillis()
  }
  
  def toc() {
    val now = System.currentTimeMillis()
    println("elapsed: " + (now-_time)/1000.0)
  }
}