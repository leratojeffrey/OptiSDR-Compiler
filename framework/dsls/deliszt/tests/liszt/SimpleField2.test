import Liszt.Language._
import Liszt.MetaInteger._


@lisztcode
object SimpleField2 {
	val ones = FieldWithConst[Cell, Int](1)
	val twos = FieldWithConst[Cell, Int](0)
	def main() {
		for (c <- cells(mesh)) {
			twos(c) += 2*ones(c)
			Print(ones(c), " ")
		}
		for (c <- cells(mesh)) {
			Print(twos(c), " ")
		}
	}
}



