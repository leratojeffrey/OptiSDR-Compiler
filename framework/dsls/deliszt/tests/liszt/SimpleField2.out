0:c1-0 liszt: 1 
0:c2-0 liszt: 1 
0:c3-0 liszt: 1 
0:c4-0 liszt: 1 
0:c5-0 liszt: 1 
0:c6-0 liszt: 1 
0:c7-0 liszt: 1 
0:c8-0 liszt: 1 
0:c9-0 liszt: 1 
0:c10-0 liszt: 1 
0:c11-0 liszt: 1 
0:c12-0 liszt: 1 
0:c13-0 liszt: 1 
0:c14-0 liszt: 1 
0:c15-0 liszt: 1 
0:c16-0 liszt: 1 
0:c17-0 liszt: 1 
0:c18-0 liszt: 1 
0:c19-0 liszt: 1 
0:c20-0 liszt: 1 
0:c21-0 liszt: 1 
0:c22-0 liszt: 1 
0:c23-0 liszt: 1 
0:c24-0 liszt: 1 
0:c25-0 liszt: 1 
0:c26-0 liszt: 1 
0:c27-0 liszt: 1 
0:c28-0 liszt: 1 
0:c29-0 liszt: 1 
0:c30-0 liszt: 1 
0:c31-0 liszt: 1 
0:c32-0 liszt: 1 
0:c33-0 liszt: 1 
0:c34-0 liszt: 1 
0:c35-0 liszt: 1 
0:c36-0 liszt: 1 
0:c37-0 liszt: 1 
0:c38-0 liszt: 1 
0:c39-0 liszt: 1 
0:c40-0 liszt: 1 
0:c41-0 liszt: 1 
0:c42-0 liszt: 1 
0:c43-0 liszt: 1 
0:c44-0 liszt: 1 
0:c45-0 liszt: 1 
0:c46-0 liszt: 1 
0:c47-0 liszt: 1 
0:c48-0 liszt: 1 
0:c49-0 liszt: 1 
0:c50-0 liszt: 1 
0:c51-0 liszt: 1 
0:c52-0 liszt: 1 
0:c53-0 liszt: 1 
0:c54-0 liszt: 1 
0:c55-0 liszt: 1 
0:c56-0 liszt: 1 
0:c57-0 liszt: 1 
0:c58-0 liszt: 1 
0:c59-0 liszt: 1 
0:c60-0 liszt: 1 
0:c61-0 liszt: 1 
0:c62-0 liszt: 1 
0:c63-0 liszt: 1 
0:c64-0 liszt: 1 
0:c65-0 liszt: 1 
0:c66-0 liszt: 1 
0:c67-0 liszt: 1 
0:c68-0 liszt: 1 
0:c69-0 liszt: 1 
0:c70-0 liszt: 1 
0:c71-0 liszt: 1 
0:c72-0 liszt: 1 
0:c73-0 liszt: 1 
0:c74-0 liszt: 1 
0:c75-0 liszt: 1 
0:c76-0 liszt: 1 
0:c77-0 liszt: 1 
0:c78-0 liszt: 1 
0:c79-0 liszt: 1 
0:c80-0 liszt: 1 
0:c81-0 liszt: 1 
0:c82-0 liszt: 1 
0:c83-0 liszt: 1 
0:c84-0 liszt: 1 
0:c85-0 liszt: 1 
0:c86-0 liszt: 1 
0:c87-0 liszt: 1 
0:c88-0 liszt: 1 
0:c89-0 liszt: 1 
0:c90-0 liszt: 1 
0:c91-0 liszt: 1 
0:c92-0 liszt: 1 
0:c93-0 liszt: 1 
0:c94-0 liszt: 1 
0:c95-0 liszt: 1 
0:c96-0 liszt: 1 
0:c97-0 liszt: 1 
0:c98-0 liszt: 1 
0:c99-0 liszt: 1 
0:c100-0 liszt: 1 
0:c101-0 liszt: 1 
0:c102-0 liszt: 1 
0:c103-0 liszt: 1 
0:c104-0 liszt: 1 
0:c105-0 liszt: 1 
0:c106-0 liszt: 1 
0:c107-0 liszt: 1 
0:c108-0 liszt: 1 
0:c109-0 liszt: 1 
0:c110-0 liszt: 1 
0:c111-0 liszt: 1 
0:c112-0 liszt: 1 
0:c113-0 liszt: 1 
0:c114-0 liszt: 1 
0:c115-0 liszt: 1 
0:c116-0 liszt: 1 
0:c117-0 liszt: 1 
0:c118-0 liszt: 1 
0:c119-0 liszt: 1 
0:c120-0 liszt: 1 
0:c121-0 liszt: 1 
0:c122-0 liszt: 1 
0:c123-0 liszt: 1 
0:c124-0 liszt: 1 
0:c125-0 liszt: 1 
0:c126-0 liszt: 1 
0:c127-0 liszt: 1 
0:c128-0 liszt: 1 
0:c129-0 liszt: 1 
0:c130-0 liszt: 1 
0:c131-0 liszt: 1 
0:c132-0 liszt: 1 
0:c133-0 liszt: 1 
0:c134-0 liszt: 1 
0:c135-0 liszt: 1 
0:c136-0 liszt: 1 
0:c137-0 liszt: 1 
0:c138-0 liszt: 1 
0:c139-0 liszt: 1 
0:c140-0 liszt: 1 
0:c141-0 liszt: 1 
0:c142-0 liszt: 1 
0:c143-0 liszt: 1 
0:c144-0 liszt: 1 
0:c145-0 liszt: 1 
0:c146-0 liszt: 1 
0:c147-0 liszt: 1 
0:c148-0 liszt: 1 
0:c149-0 liszt: 1 
0:c150-0 liszt: 1 
0:c151-0 liszt: 1 
0:c152-0 liszt: 1 
0:c153-0 liszt: 1 
0:c154-0 liszt: 1 
0:c155-0 liszt: 1 
0:c156-0 liszt: 1 
0:c157-0 liszt: 1 
0:c158-0 liszt: 1 
0:c159-0 liszt: 1 
0:c160-0 liszt: 1 
0:c161-0 liszt: 1 
0:c162-0 liszt: 1 
0:c163-0 liszt: 1 
0:c164-0 liszt: 1 
0:c165-0 liszt: 1 
0:c166-0 liszt: 1 
0:c167-0 liszt: 1 
0:c168-0 liszt: 1 
0:c169-0 liszt: 1 
0:c170-0 liszt: 1 
0:c171-0 liszt: 1 
0:c172-0 liszt: 1 
0:c173-0 liszt: 1 
0:c174-0 liszt: 1 
0:c175-0 liszt: 1 
0:c176-0 liszt: 1 
0:c177-0 liszt: 1 
0:c178-0 liszt: 1 
0:c179-0 liszt: 1 
0:c180-0 liszt: 1 
0:c181-0 liszt: 1 
0:c182-0 liszt: 1 
0:c183-0 liszt: 1 
0:c184-0 liszt: 1 
0:c185-0 liszt: 1 
0:c186-0 liszt: 1 
0:c187-0 liszt: 1 
0:c188-0 liszt: 1 
0:c189-0 liszt: 1 
0:c190-0 liszt: 1 
0:c191-0 liszt: 1 
0:c192-0 liszt: 1 
0:c193-0 liszt: 1 
0:c194-0 liszt: 1 
0:c195-0 liszt: 1 
0:c196-0 liszt: 1 
0:c197-0 liszt: 1 
0:c198-0 liszt: 1 
0:c199-0 liszt: 1 
0:c200-0 liszt: 1 
0:c201-0 liszt: 1 
0:c202-0 liszt: 1 
0:c203-0 liszt: 1 
0:c204-0 liszt: 1 
0:c205-0 liszt: 1 
0:c206-0 liszt: 1 
0:c207-0 liszt: 1 
0:c208-0 liszt: 1 
0:c209-0 liszt: 1 
0:c210-0 liszt: 1 
0:c211-0 liszt: 1 
0:c212-0 liszt: 1 
0:c213-0 liszt: 1 
0:c214-0 liszt: 1 
0:c215-0 liszt: 1 
0:c216-0 liszt: 1 
0:c217-0 liszt: 1 
0:c218-0 liszt: 1 
0:c219-0 liszt: 1 
0:c220-0 liszt: 1 
0:c221-0 liszt: 1 
0:c222-0 liszt: 1 
0:c223-0 liszt: 1 
0:c224-0 liszt: 1 
0:c225-0 liszt: 1 
0:c226-0 liszt: 1 
0:c227-0 liszt: 1 
0:c228-0 liszt: 1 
0:c229-0 liszt: 1 
0:c230-0 liszt: 1 
0:c231-0 liszt: 1 
0:c232-0 liszt: 1 
0:c233-0 liszt: 1 
0:c234-0 liszt: 1 
0:c235-0 liszt: 1 
0:c236-0 liszt: 1 
0:c237-0 liszt: 1 
0:c238-0 liszt: 1 
0:c239-0 liszt: 1 
0:c240-0 liszt: 1 
0:c241-0 liszt: 1 
0:c242-0 liszt: 1 
0:c243-0 liszt: 1 
0:c244-0 liszt: 1 
0:c245-0 liszt: 1 
0:c246-0 liszt: 1 
0:c247-0 liszt: 1 
0:c248-0 liszt: 1 
0:c249-0 liszt: 1 
0:c250-0 liszt: 1 
1:c1-0 liszt: 2 
1:c2-0 liszt: 2 
1:c3-0 liszt: 2 
1:c4-0 liszt: 2 
1:c5-0 liszt: 2 
1:c6-0 liszt: 2 
1:c7-0 liszt: 2 
1:c8-0 liszt: 2 
1:c9-0 liszt: 2 
1:c10-0 liszt: 2 
1:c11-0 liszt: 2 
1:c12-0 liszt: 2 
1:c13-0 liszt: 2 
1:c14-0 liszt: 2 
1:c15-0 liszt: 2 
1:c16-0 liszt: 2 
1:c17-0 liszt: 2 
1:c18-0 liszt: 2 
1:c19-0 liszt: 2 
1:c20-0 liszt: 2 
1:c21-0 liszt: 2 
1:c22-0 liszt: 2 
1:c23-0 liszt: 2 
1:c24-0 liszt: 2 
1:c25-0 liszt: 2 
1:c26-0 liszt: 2 
1:c27-0 liszt: 2 
1:c28-0 liszt: 2 
1:c29-0 liszt: 2 
1:c30-0 liszt: 2 
1:c31-0 liszt: 2 
1:c32-0 liszt: 2 
1:c33-0 liszt: 2 
1:c34-0 liszt: 2 
1:c35-0 liszt: 2 
1:c36-0 liszt: 2 
1:c37-0 liszt: 2 
1:c38-0 liszt: 2 
1:c39-0 liszt: 2 
1:c40-0 liszt: 2 
1:c41-0 liszt: 2 
1:c42-0 liszt: 2 
1:c43-0 liszt: 2 
1:c44-0 liszt: 2 
1:c45-0 liszt: 2 
1:c46-0 liszt: 2 
1:c47-0 liszt: 2 
1:c48-0 liszt: 2 
1:c49-0 liszt: 2 
1:c50-0 liszt: 2 
1:c51-0 liszt: 2 
1:c52-0 liszt: 2 
1:c53-0 liszt: 2 
1:c54-0 liszt: 2 
1:c55-0 liszt: 2 
1:c56-0 liszt: 2 
1:c57-0 liszt: 2 
1:c58-0 liszt: 2 
1:c59-0 liszt: 2 
1:c60-0 liszt: 2 
1:c61-0 liszt: 2 
1:c62-0 liszt: 2 
1:c63-0 liszt: 2 
1:c64-0 liszt: 2 
1:c65-0 liszt: 2 
1:c66-0 liszt: 2 
1:c67-0 liszt: 2 
1:c68-0 liszt: 2 
1:c69-0 liszt: 2 
1:c70-0 liszt: 2 
1:c71-0 liszt: 2 
1:c72-0 liszt: 2 
1:c73-0 liszt: 2 
1:c74-0 liszt: 2 
1:c75-0 liszt: 2 
1:c76-0 liszt: 2 
1:c77-0 liszt: 2 
1:c78-0 liszt: 2 
1:c79-0 liszt: 2 
1:c80-0 liszt: 2 
1:c81-0 liszt: 2 
1:c82-0 liszt: 2 
1:c83-0 liszt: 2 
1:c84-0 liszt: 2 
1:c85-0 liszt: 2 
1:c86-0 liszt: 2 
1:c87-0 liszt: 2 
1:c88-0 liszt: 2 
1:c89-0 liszt: 2 
1:c90-0 liszt: 2 
1:c91-0 liszt: 2 
1:c92-0 liszt: 2 
1:c93-0 liszt: 2 
1:c94-0 liszt: 2 
1:c95-0 liszt: 2 
1:c96-0 liszt: 2 
1:c97-0 liszt: 2 
1:c98-0 liszt: 2 
1:c99-0 liszt: 2 
1:c100-0 liszt: 2 
1:c101-0 liszt: 2 
1:c102-0 liszt: 2 
1:c103-0 liszt: 2 
1:c104-0 liszt: 2 
1:c105-0 liszt: 2 
1:c106-0 liszt: 2 
1:c107-0 liszt: 2 
1:c108-0 liszt: 2 
1:c109-0 liszt: 2 
1:c110-0 liszt: 2 
1:c111-0 liszt: 2 
1:c112-0 liszt: 2 
1:c113-0 liszt: 2 
1:c114-0 liszt: 2 
1:c115-0 liszt: 2 
1:c116-0 liszt: 2 
1:c117-0 liszt: 2 
1:c118-0 liszt: 2 
1:c119-0 liszt: 2 
1:c120-0 liszt: 2 
1:c121-0 liszt: 2 
1:c122-0 liszt: 2 
1:c123-0 liszt: 2 
1:c124-0 liszt: 2 
1:c125-0 liszt: 2 
1:c126-0 liszt: 2 
1:c127-0 liszt: 2 
1:c128-0 liszt: 2 
1:c129-0 liszt: 2 
1:c130-0 liszt: 2 
1:c131-0 liszt: 2 
1:c132-0 liszt: 2 
1:c133-0 liszt: 2 
1:c134-0 liszt: 2 
1:c135-0 liszt: 2 
1:c136-0 liszt: 2 
1:c137-0 liszt: 2 
1:c138-0 liszt: 2 
1:c139-0 liszt: 2 
1:c140-0 liszt: 2 
1:c141-0 liszt: 2 
1:c142-0 liszt: 2 
1:c143-0 liszt: 2 
1:c144-0 liszt: 2 
1:c145-0 liszt: 2 
1:c146-0 liszt: 2 
1:c147-0 liszt: 2 
1:c148-0 liszt: 2 
1:c149-0 liszt: 2 
1:c150-0 liszt: 2 
1:c151-0 liszt: 2 
1:c152-0 liszt: 2 
1:c153-0 liszt: 2 
1:c154-0 liszt: 2 
1:c155-0 liszt: 2 
1:c156-0 liszt: 2 
1:c157-0 liszt: 2 
1:c158-0 liszt: 2 
1:c159-0 liszt: 2 
1:c160-0 liszt: 2 
1:c161-0 liszt: 2 
1:c162-0 liszt: 2 
1:c163-0 liszt: 2 
1:c164-0 liszt: 2 
1:c165-0 liszt: 2 
1:c166-0 liszt: 2 
1:c167-0 liszt: 2 
1:c168-0 liszt: 2 
1:c169-0 liszt: 2 
1:c170-0 liszt: 2 
1:c171-0 liszt: 2 
1:c172-0 liszt: 2 
1:c173-0 liszt: 2 
1:c174-0 liszt: 2 
1:c175-0 liszt: 2 
1:c176-0 liszt: 2 
1:c177-0 liszt: 2 
1:c178-0 liszt: 2 
1:c179-0 liszt: 2 
1:c180-0 liszt: 2 
1:c181-0 liszt: 2 
1:c182-0 liszt: 2 
1:c183-0 liszt: 2 
1:c184-0 liszt: 2 
1:c185-0 liszt: 2 
1:c186-0 liszt: 2 
1:c187-0 liszt: 2 
1:c188-0 liszt: 2 
1:c189-0 liszt: 2 
1:c190-0 liszt: 2 
1:c191-0 liszt: 2 
1:c192-0 liszt: 2 
1:c193-0 liszt: 2 
1:c194-0 liszt: 2 
1:c195-0 liszt: 2 
1:c196-0 liszt: 2 
1:c197-0 liszt: 2 
1:c198-0 liszt: 2 
1:c199-0 liszt: 2 
1:c200-0 liszt: 2 
1:c201-0 liszt: 2 
1:c202-0 liszt: 2 
1:c203-0 liszt: 2 
1:c204-0 liszt: 2 
1:c205-0 liszt: 2 
1:c206-0 liszt: 2 
1:c207-0 liszt: 2 
1:c208-0 liszt: 2 
1:c209-0 liszt: 2 
1:c210-0 liszt: 2 
1:c211-0 liszt: 2 
1:c212-0 liszt: 2 
1:c213-0 liszt: 2 
1:c214-0 liszt: 2 
1:c215-0 liszt: 2 
1:c216-0 liszt: 2 
1:c217-0 liszt: 2 
1:c218-0 liszt: 2 
1:c219-0 liszt: 2 
1:c220-0 liszt: 2 
1:c221-0 liszt: 2 
1:c222-0 liszt: 2 
1:c223-0 liszt: 2 
1:c224-0 liszt: 2 
1:c225-0 liszt: 2 
1:c226-0 liszt: 2 
1:c227-0 liszt: 2 
1:c228-0 liszt: 2 
1:c229-0 liszt: 2 
1:c230-0 liszt: 2 
1:c231-0 liszt: 2 
1:c232-0 liszt: 2 
1:c233-0 liszt: 2 
1:c234-0 liszt: 2 
1:c235-0 liszt: 2 
1:c236-0 liszt: 2 
1:c237-0 liszt: 2 
1:c238-0 liszt: 2 
1:c239-0 liszt: 2 
1:c240-0 liszt: 2 
1:c241-0 liszt: 2 
1:c242-0 liszt: 2 
1:c243-0 liszt: 2 
1:c244-0 liszt: 2 
1:c245-0 liszt: 2 
1:c246-0 liszt: 2 
1:c247-0 liszt: 2 
1:c248-0 liszt: 2 
1:c249-0 liszt: 2 
1:c250-0 liszt: 2 
